File: packages/agent/src/session-context.ts
import { Buffer } from "node:buffer";
import { MEMORY_KEYS, type MemoryStore } from "./db/memory";

// A session owns its environment and working memory. Prompt rendering is a
// separate operation: renderContext includes the small environment block,
// while stored tool results are passed to consumers explicitly.

// Narrow dependency surface: gathering env data needs one MCP call, one
// memory read and the user's email — not the whole Engine. `userEmail` is
// read from env ONCE in the composition root and injected here, so this
// per-signal business path touches no process.env.
export interface EnvDataDeps {
  mcp: { callTool(name: string, args: Record<string, unknown>): Promise<string> };
  memory: Pick<MemoryStore, "get">;
  userEmail: string | null;
}

// Structured env data — single source of truth for both the supervisor
// (markdown context block) and the workflow runner (variable store
// initial value under the `env` key). When this shape changes, both
// consumers update at once.
export interface EnvData {
  now: Date;
  timezone: string;
  userEmail: string | null;
  newsLastReadAt: string | null;
}

// A caller-supplied label. The store neither parses nor validates JSON.
export type WorkingMemoryFormat = "text" | "json";

export interface WorkingMemoryEntry {
  key: string;
  format: WorkingMemoryFormat;
  sizeBytes: number;
}

export interface WorkingMemory {
  // Insert only: keys are exact, non-empty strings. An occupied key throws.
  put(key: string, value: string, format?: WorkingMemoryFormat): void;
  // Returns the original string, including empty strings. Missing keys throw.
  get(key: string): string;
  // Detached metadata in insertion order; payloads stay out of the listing.
  list(): WorkingMemoryEntry[];
  // True if removed, false if absent. The deleted key may then be reused.
  delete(key: string): boolean;
}

export interface SessionContext {
  readonly id: string;
  readonly env: EnvData;
  readonly memory: WorkingMemory;
}

interface StoredValue {
  value: string;
  format: WorkingMemoryFormat;
  sizeBytes: number;
}

// Each call creates independent memory. Callers share the context explicitly
// with the components working on the same task and control its lifetime.
export function createSessionContext({ id, env }: { id: string; env: EnvData }): SessionContext {
  const entries = new Map<string, StoredValue>();

  return {
    id,
    env,
    memory: {
      put(key, value, format = "text") {
        if (key.length === 0) throw new Error("Key must not be empty");
        if (entries.has(key)) throw new Error(`Key ${JSON.stringify(key)} already exists`);

        entries.set(key, { value, format, sizeBytes: Buffer.byteLength(value, "utf8") });
      },

      get(key) {
        const entry = entries.get(key);
        if (entry === undefined) throw new Error(`Key ${JSON.stringify(key)} not found`);
        return entry.value;
      },

      list() {
        return Array.from(entries, ([key, { format, sizeBytes }]) => ({
          key,
          format,
          sizeBytes,
        }));
      },

      delete(key) {
        return entries.delete(key);
      },
    },
  };
}

function formatLocalTime(now: Date, tz: string): string {
  const parts = new Intl.DateTimeFormat("en-CA", {
    timeZone: tz,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    hour12: false,
  }).formatToParts(now);
  const get = (type: string) => parts.find((p) => p.type === type)?.value ?? "";
  return `${get("year")}-${get("month")}-${get("day")} ${get("hour")}:${get("minute")}`;
}

// Reads the integration-owned timezone from MCP exactly once per call.
// Returns "UTC" if the MCP call fails — the block is best-effort, we'd
// rather inject a slightly-wrong tz than crash the session.
async function readTimezone(deps: EnvDataDeps): Promise<string> {
  try {
    const raw = await deps.mcp.callTool("get_timezone", {});
    if (raw.startsWith("[tool error]")) return "UTC";
    const parsed = JSON.parse(raw) as { timezone?: string };
    return parsed.timezone ?? "UTC";
  } catch {
    return "UTC";
  }
}

export async function gatherEnvData(deps: EnvDataDeps): Promise<EnvData> {
  const tz = await readTimezone(deps);
  return {
    now: new Date(),
    timezone: tz,
    userEmail: deps.userEmail,
    newsLastReadAt: deps.memory.get(MEMORY_KEYS.newsLastReadAt),
  };
}

// Pure render of the already-gathered EnvData into the markdown block.
export function renderContext(env: EnvData): string {
  const lines = [
    "## Current context",
    `- Local time: ${formatLocalTime(env.now, env.timezone)} (${env.timezone})`,
  ];
  if (env.userEmail) lines.push(`- User email: ${env.userEmail}`);
  lines.push(
    `- News last read at: ${env.newsLastReadAt ?? "never (bootstrap with now - 24h)"}`,
  );
  return lines.join("\n");
}

File: agent-py/agent_py/session_context.py
"""The ``Current context`` block + structured env data.

A small block prepended to every session's system prompt. It saves the agent a
couple of tool calls (no need to ask for timezone / current time) and gives
every skill one canonical "state of the world".
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from datetime import UTC, datetime
from zoneinfo import ZoneInfo

from .db.memory import MEMORY_KEYS, get_memory


@dataclass
class EnvData:
    """Structured env data — the single source of truth for both the supervisor
    (the markdown block) and the workflow engine (initial values of the variable
    store under the ``env`` key)."""

    now: datetime
    timezone: str
    user_email: str | None
    news_last_read_at: str | None


def _format_local_time(now: datetime, tz: str) -> str:
    try:
        local = now.astimezone(ZoneInfo(tz))
    except Exception:  # noqa: BLE001 — unknown tz name; fall back to UTC
        local = now.astimezone(UTC)
    return local.strftime("%Y-%m-%d %H:%M")


def gather_env_data(tz: str) -> EnvData:
    """Gather env data. ``tz`` is fetched from MCP by the engine beforehand (an
    async call) so this module has no direct dependency on the MCP client."""
    return EnvData(
        now=datetime.now(UTC),
        timezone=tz,
        user_email=os.environ.get("USER_EMAIL"),
        news_last_read_at=get_memory(MEMORY_KEYS["news_last_read_at"]),
    )


def build_session_context(env: EnvData) -> str:
    lines = [
        "## Current context",
        f"- Local time: {_format_local_time(env.now, env.timezone)} ({env.timezone})",
    ]
    if env.user_email:
        lines.append(f"- User email: {env.user_email}")
    lines.append(
        f"- News last read at: {env.news_last_read_at or 'never (bootstrap with now - 24h)'}"
    )
    return "\n".join(lines)
